qiskit/circuit/singleton.py

```diff
@@ -0,0 +1,250 @@
+
+    def __init_subclass__(cls, *, create_singleton=True, **kwargs):
+        super(base, cls).__init_subclass__(**kwargs)
+        if not create_singleton:
+            return
+
+        # We need to make a new type object that pokes in the overrides into the correct
+        # place in the method-resolution order.
+        singleton_class = _SingletonMeta.__new__(
+            _SingletonMeta,
+            f"_Singleton{cls.__name__}",
+            (cls, overrides),
+            # This is a dynamically generated class so it's got no module.  The slot layout of the
+            # singleton class needs to match any layout in the base.
+            {"__module__": None, "__slots__": (), "__new__": _impl_new, "_base_class": cls},
+            create_singleton=False,
+        )
+
+        # Make a mutable instance, fully instantiate all lazy properties, then freeze it.
+        cls._singleton_instance = cls(_force_mutable=True)
+        cls._singleton_instance._define()
+        cls._singleton_instance.__class__ = singleton_class
+
+    return classmethod(__init_subclass__)
+
+
+class _SingletonMeta(type(Instruction)):
+    # The inheritance above is to ensure metaclass compatibility with `Instruction`, though pylint
+    # doesn't understand that this is a subclass of `type`, so uses metaclass `self` conventions.
+
+    # pylint: disable=bad-classmethod-argument,no-self-argument
+
+    __slots__ = ()
+
+    def __new__(mcs, name, bases, namespace, *, overrides=None, **kwargs):
+        cls = super().__new__(mcs, name, bases, namespace, **kwargs)
+        if overrides is not None:
+            # The `__init_subclass__` logic is shared between `SingletonInstruction` and
+            # `SingletonGate`, but we can't do `__init_subclass__ = _impl_init_subclass(...)`
+            # inside the class definition bodies because it couldn't make `super()` calls.
+            cls.__init_subclass__ = _impl_init_subclass(cls, overrides)
+        return cls
+
+    def __call__(cls, *args, _force_mutable=False, **kwargs):
+        if not _force_mutable and not args and not kwargs:
+            # This class attribute is created by the singleton-creation base classes'
+            # `__init_subclass__` methods; see `_impl_init_subclass`.
+            return cls._singleton_instance
+        return super().__call__(*args, **kwargs)
+
+
+class _SingletonInstructionOverrides(Instruction):
+    """Overrides for all the mutable methods and properties of `Instruction` to make it
+    immutable."""
+
+    __slots__ = ()
+
+    def c_if(self, classical, val):
+        return self.to_mutable().c_if(classical, val)
+
+    @property
+    def base_class(self):
+        # `type(self)` will actually be the dynamic `_SingletonXGate` (e.g.) created by
+        # `SingletonGate.__init_subclass__` during the instantiation of `XGate`, since this class
+        # is never the concrete type of a class.
+        return type(self)._base_class
```

Comment:
I was thinking maybe we could override `__class__` to return `_base_class` but I don't think that's foolproof in every scenario so this might be the best solution.

Reply:
Making `__class__` a property wouldn't trick `type(obj)` which is the main place we need it to work, and if we statically assigned to it we'd make the singleton mutable again.